Code/GraphMol/MolStandardize/Tautomer.h

```diff
@@ -40,29 +43,247 @@ inline int scoreTautomer(const ROMol &mol) {
 }
 }  // namespace TautomerScoringFunctions
 
+enum TautomerEnumeratorStatus {
+  Completed = 0,
+  MaxTautomersReached,
+  MaxTransformsReached,
+  Canceled
+};
+
+class Tautomer {
+  friend class TautomerEnumerator;
+
+ public:
+  Tautomer() : d_numModifiedAtoms(0), d_numModifiedBonds(0) {}
+  Tautomer(const ROMOL_SPTR &t, const ROMOL_SPTR &k, size_t a = 0, size_t b = 0)
+      : tautomer(t),
+        kekulized(k),
+        d_numModifiedAtoms(a),
+        d_numModifiedBonds(b) {}
+  ROMOL_SPTR tautomer;
+  ROMOL_SPTR kekulized;
+
+ private:
+  size_t d_numModifiedAtoms;
+  size_t d_numModifiedBonds;
+};
+
+typedef std::map<std::string, Tautomer> SmilesTautomerMap;
+typedef std::pair<std::string, Tautomer> SmilesTautomerPair;
+
+//! Contains results of tautomer enumeration
+class RDKIT_MOLSTANDARDIZE_EXPORT TautomerEnumeratorResult {
+  friend class TautomerEnumerator;
+
+ public:
+  class const_iterator {
+   public:
+    typedef ROMOL_SPTR value_type;
+    typedef std::ptrdiff_t difference_type;
+    typedef const ROMOL_SPTR *pointer;
+    typedef const ROMOL_SPTR &reference;
+    typedef std::input_iterator_tag iterator_category;
+
+    explicit const_iterator(const SmilesTautomerMap::const_iterator &it)
+        : d_it(it) {}
+    reference operator*() const { return d_it->second.tautomer; }
+    bool operator==(const const_iterator &other) const {
+      return (d_it == other.d_it);
+    }
+    bool operator!=(const const_iterator &other) const {
+      return !(*this == other);
+    }
+    const_iterator operator++(int) {
+      const_iterator copy(d_it);
+      ++*this;
```

Comment:
This doesn't look to do what is advertised.  In fact It is unclear that is happening.  You are returning the unincremented copy, but advancing the internal copy (and not returning it).

It appears that SmilesTautomerMap+=n is the same as SmilesTautomerMap+=1 which appears to be different than SmilesTautomerMap++;

Reply:
If we really need random access, perhaps std::next is what you want.

If we don't need random access, I think instead of the const_iterator class we could just return

```
const SmilesTautomerMap::const_iterator TautomerEnumeratorResult::begin() {
  return d_tautomers.begin();
}
```

Reply:
I am pretty sure it does what is advertised.
`const_iterator &operator++()` is the standard pre-increment operator, which increments the iterator and returns itself.
`const_iterator operator++(int)` is the standard post-increment operator, The dummy `int` argument is needed to differentiate the post-increment operator from the pre-increment one, but it is not actually used: https://en.cppreference.com/w/cpp/language/operator_incdec
The post-increment operator makes a copy of `this` first, as the iterator should not change in the returned copy.
Then, it calls the pre-increment operator on `this`, which actually increments the iterator.
Finally, it returns the copy with the original iterator. The net result is that the iterator is increment, but the non-incremented version of the iterator is returned, which is the expected behavior of the post-increment operator.
I cannot see anything wrong with it.

The `+=` operator is not supported at all, as this is not a random access iterator.
It used to be a simple `input_iterator`; as the underlying `std::map::const_iterator` supports bidirectional iterators, I added the pre- and post-decrement operator and made it a `bidirectional_iterator`, and added relevant test to prove that it does what it is supposed to do.

We cannot simply return `d_tautomers.begin()` as then we would be iterating on key, value pairs, where key is `std::string` and value is `Tautomer`. Instead, to maintain backwards compatibility we need to iterate on `ROMOL_SPTR`.

I hope the above makes sense to you.

Reply:
@ptosco - ah I had forgotten that ++(int) returns the current value and then applies the postfix.  The (int) was to differentiate this from the postfix with no return.  Obvious, eh? :)

Reply:
What I'm confused by is that ++(int) essentially does post fix on:

```
++(*this);
```
which is
```
++d_it->second.tautomer
```

and ++() does
```
++d_it;
```

which are not the same thing.

Reply:
Nope, that's not what happens.
Doing:
```
++*this;
```
is equivalent to doing:
```
operator++();
```
i.e., the post-increment operator calls the pre-increment one after having copied itself in order to return a non-incremented copy of itself.
You can see that very clearly adding some debugging `std::cerr`s:
```
    const_iterator operator++(int) {
      const_iterator copy(d_it);
      std::cerr << "I am the post-increment operator and now I'll call my pre-increment mate" << std::endl;
      ++*this;
      std::cerr << "I am the post-increment operator and I have just called my pre-increment mate" << std::endl;
      return copy;
    }
    const_iterator &operator++() {
      std::cerr << "I am the pre-increment operator" << std::endl;
      ++d_it;
      return *this;
    }
```
Then when the post-increment operator is called you see the following:
```
1) testTautomerEnumeratorResult_const_iterator calling post-increment
I am the post-increment operator and now I'll call my pre-increment mate
I am the pre-increment operator
I am the post-increment operator and I have just called my pre-increment mate
```

Reply:
Got it, I was reading

```
 ++*(*this)
```

not

```
 ++(*this)
```

That's some confusing syntax.  Perhaps

```
this->operator++();
```

Would be easier to read.

Reply:
In any case, sorry about my slow comprehension here.

Reply:
@bp-kelley I have applied your suggested change. I guess we are good to merge once tests have passed.